tempesta_fw/ss_skb.c

```diff
@@ -1067,6 +1067,7 @@ ss_skb_process(struct sk_buff *skb, unsigned int off, unsigned int trail,
 			n = frag_size - off;
 			if (trail > frag_len)
 				n -= trail - frag_len;
+			_processed = 0;
 			r = actor(objdata, frag_addr + off, n, &_processed);
 			*processed += _processed;
```

Comment:
Good catch - `*processed += _processed` really seems very wrong. But `ttls_recv()` and probably other actors use `_processed` as a counter of currently read record, so it seems wrong to zeroize it.

Reply:
Yeah, that part is tricky. I'll look further into it.

Reply:
Throughout the code `ss_skb_process()` is called with various "actor" callbacks. Some of them increment `*parsed`, some do not. Specifically:
* ttls_recv — increments;
* tfw_http_parse_req — resets to `0` immediately, and then writes a new value;
* tfw_http_parse_resp — resets to `0`, then writes a new value;
* tfw_h2_frame_recv — increments;
* tfw_bmb_print_msg — writes a new value that discards any previous value.

I haven't found any dependency on current values of `*parsed` in functions. There are some uses, but those save values to figure out relative changes of the counter after child function calls. From what I found I think it's safe to reset `_processed` before calling `actor()`. Effectively, http parsing routines do reset the counter. Only TLS-related do not.

Reply:
Yes, agree. Please also note that `ttls_parse_client_hello()` doesn't even update `read` - looks like a bug. Please also check other handshake functions against `read` usage.

Reply:
> `ttls_parse_client_hello()` doesn't even update read

There is an update hidden in `TTLS_HS_FSM_FINISH()`: `*read += p - buf;`.